dali/image/transform.cc

```diff
@@ -152,4 +152,47 @@ DALIError_t FastResizeCropMirrorHost(const uint8 *img, int H, int W, int C,
   return DALISuccess;
 }
 
+void CheckParam(const Tensor<CPUBackend> &input, const std::string &opName) {
+  DALI_ENFORCE(input.ndim() == 3);
+  DALI_ENFORCE(IsType<uint8>(input.type()),
+               opName + " expects input data in uint8.");
+  DALI_ENFORCE(input.dim(2) == 1 || input.dim(2) == 3,
+               opName + " supports hwc rgb & grayscale inputs.");
+}
+
+typedef cv::Vec<uchar, 1> Vec1b;
+
+DALIError_t MakeColorTransformation(const uint8 *img, int H, int W, int C,
+                                    const float *matrix, uint8 *out_img) {
+  const int channel_flag = C == 3 ? CV_8UC3 : CV_8UC1;
+
+  const cv::Mat cv_imgIn = CreateMatFromPtr(H, W, channel_flag, img);
+  cv::Mat cv_imgOut = CreateMatFromPtr(H, W, channel_flag, out_img);
+
+  if (C == 1) {
+    for (int y = 0; y < H; ++y) {
+      for (int x = 0; x < W; ++x) {
+        cv_imgOut.at<Vec1b>(y, x)[0] =
+            cv::saturate_cast<uint8>((matrix[0] * cv_imgIn.at<Vec1b>(y, x)[0]) + matrix[1]);
+      }
+    }
+  } else {
+    for (int y = 0; y < H; ++y) {
+      for (int x = 0; x < W; ++x) {
+        const auto &inpPix = cv_imgIn.at<cv::Vec3b>(y, x);
+        auto matrRow = matrix;
+        for (int k = 0; k < C; ++k, matrRow += 4) {
+          auto val = matrRow[3];
+          for (int n = 0; n < C; ++n)
+            val += inpPix[n] * matrRow[n];
+
```

Comment:
So if I'm reading this right, the transformation matrix is 4x4, but the input is 3 channels, so we just treat it as if alpha==1.f.  Right?  (Are 1 and 3 the only valid #'s of input channels?  Is that enforced anywhere?)

I'm basically thinking these loops are so short that just manually unrolling them might be both more clear and possibly faster.

Reply:
1. Yes, today the color transformations on GPU are implemented for C = 1 and 3. 

2. As for transparency - we don't use is, but everything is prepared for it. We construct 
4x4 matrices, but, as we could see in `color_twist.cu`
```
typedef NppStatus (*colorTwistFunc)(const Npp8u *pSrc, int nSrcStep, Npp8u *pDst, 
int nDstStep, NppiSize oSizeROI, const Npp32f aTwist[3][4]);
```
we are using only 3 first rows just because this is standard input for   `nppiColorTwist32f_8u_C3R : nppiColorTwist32f_8u_C1R` methods.

Similar approach was used for CPU.

3. I am not sure if direct calculation will be significantly faster. I will do some experiments.

Reply:
My experiment shows that following code
```
for (int y = 0; y < H; ++y) {
   for (int x = 0; x < W; ++x) {
      const auto &inpPix = cv_imgIn.at<cv::Vec3b>(y, x);
      auto &outPix = cv_imgOut.at<cv::Vec3b>(y, x);
      outPix[0] = cv::saturate_cast<uint8>
           (matrRow[3] + inpPix[0] * matrRow[0] + inpPix[1] * matrRow[1] + inpPix[2] * matrRow[2]);
      outPix[1] = cv::saturate_cast<uint8>
           (matrRow[7] + inpPix[0] * matrRow[4] + inpPix[1] * matrRow[5] + inpPix[2] * matrRow[6]);
      outPix[2] = cv::saturate_cast<uint8>
         (matrRow[11] + inpPix[0] * matrRow[8] + inpPix[1] * matrRow[9] + inpPix[2] * matrRow[10]);
    }
}
```
is approx. **25% faster** than previously used

```
for (int y = 0; y < H; ++y) {
   for (int x = 0; x < W; ++x) {
      const auto &inpPix = cv_imgIn.at<cv::Vec3b>(y, x);
      auto matrRow = matrix;
       for (int k = 0; k < C; ++k, matrRow += 4) {
          auto val = matrRow[3];
           for (int n = 0; n < C; ++n)
               val += inpPix[n] * matrRow[n];

          cv_imgOut.at<cv::Vec3b>(y, x)[k] = cv::saturate_cast<uint8>(val);
       }
    }
}
```
I was testing one thread running 100 iterations of the batch with 128 images.  The program spends  6.36 sec. vs 8.38 sec in that function. 

It looks like it makes sense to replace two internal loops by direct calculations.